### DISCUSSION TITLE
Expected behaviour when including multiple Where() filters

### DISCUSSION THREAD

**Discussion by TheCakeMonster** [DotNotStandard/DependencyInjection#5] [Q&A]
@swegele As my first user, can I ask you a hypothetical question? It's about what happens under more advanced filtering scenarios, with multiple Where() tests. It popped into my mind after looking at your lambda in #3 

Given the following code in an assembly containing the 3 types FakeBusinessFactory, MainBusinessFactory and FakeRepository, what would be your expectation?

```csharp
services.DiscoverTypes()
	.Where(type => type.Name.StartsWith("Fake"))
	.Where(type => type.Name.EndsWith("Factory"))
	.AsThemselves()
	.Register();
``` 
How many classes would you expect to be registered as a result of this being run at startup?

**Comment by swegele** [comment]
Honestly I think without adding something more, you'd end up with a "Tom-A-to" vs "Tom-AH-to" scenario where it seems correct for only half the people.

What comes to mind is SQL where clauses with AND, OR etc.  [Here is a link](https://www.llblgen.com/Documentation/5.9/LLBLGen%20Pro%20RTF/Using%20the%20generated%20code/Filtering%20and%20Sorting/gencode_filteringpredicateclasses.htm#constructing-predicate-expressions) to my goto ORM framework where he (Frans Bouma) added those keywords to his framework to allow mimicing TSQL.
Maybe you could add something like that.

He defaults to "AND" if the clause is not explicitly specified.

**Reply by TheCakeMonster** [reply]
Yeah, this is certainly the danger - that I can only please half of the people. 

This is a feature that already exists, not anything that I am adding in extra. I made a choice and I'm now wondering whether that choice was the right one. If I'm going to change it, I'd best do so before other people use it.

Are you saying that you think AND is the best choice if it's one or the other?

**Reply by swegele** [reply]
Well if your not going to make it explicit by adding And/Or clauses....then IMHO you should either:
- throw an error if they try to do multiple WHERE
- or last in wins - only take the last WHERE

Makes me wonder if there is a kind of norm in the community when encountering this...I'd think it's fairly common problem with Fluent pattern.  
If all else fails, remember what we learned in school when second guessing ourselves on tests...always go with your first instinct :-)
